### DISCUSSION TITLE
"Load more comments" in issues and PRs needs to work better

### DISCUSSION THREAD

**Discussion by LoneFenris** [community/community#5119] [Projects and Issues]
The way Github collapses issue and PR comments based on number is rather poor. When clicking the Load More link, Github may still not load all comments for the issue and the "Load more" link jumps down below the additional comments. This requires a user searching the entire issue to repeatedly "Load more", scroll down, "Load more", scroll down, etc. This is liable to be a common user desire as the "Load more" link loads old comments first -- requiring a user that is looking for the most recent hidden comments to need to expand the full conversation.

There should be a way for a user to easily fully load an issue's thread of conversation. This could be in the form of a "Load all comments" link, or at the least having the page follow the end of the newly loaded comments when clicking "Load more" so that a user may repeatedly click the link.

**Comment by iandunn** [comment]
I agree, all of the "Show hidden comments" type are very frustrating. I often have to click a dozen extra times, instead of just scrolling or `page down`.

A single button to show all the hidden content would be nice. Even better would be a user option to just never hide things in the first place.

**Reply by iandunn** [reply]
>  it's bad idea for massive threads like 500-2500 comments

That's fair, maybe instead of on/off, the option could let each user set their preference of when to start hiding comments.

**Reply by LoneFenris** [reply]
> while it might be good for threads with less than 500 comments, it's bad idea for massive threads like 500-2500 comments, see why: "Load more comments" in issues and PRs needs to work better #5119 (comment), also it's currently possible to unhide all comments with userscripts if you need

Perhaps such extreme cases need a way to switch to a paginated view? Or perhaps a more intelligent/manual way of loading comment content similar to how sites like reddit work where you can click on the comments you want to read.

I wonder, though, if the performance issue you cite is the result of using the userscript you mentioned. If so, I then wonder if the userscript was the source of the performance problem -- trying to programmatically interact with "Load more comments" on the site to get them all loaded and running a separate comments request for each one. If that's the case, having a single "all comments" request option would likely solve that problem.

I do understand there is use case justification for partial loading from the end as well, though.

**Reply by iandunn** [reply]
What % of threads have thousands of comments?

That seems like hurting the normal case to support an edge case. The suggestions here also include options for users to turn it on/off, or switching to a paginated view after `n` comments are reached. `n` could be configured per-repo and/or user, with a reasonable default like `75`.

**Reply by garry-ut99** [reply]
> LoneFenris: Perhaps such extreme cases need a way to switch to a paginated view? Or perhaps a more intelligent/manual way of loading comment content similar to how sites like reddit work where you can click on the comments you want to read.

Yes, or this smart solution: https://github.com/orgs/community/discussions/5119#discussioncomment-15431886

<br>

> iandunn: What % of threads have thousands of comments?
> iandunn: That seems like hurting the normal case to support an edge case. 

Using percentages as a measure is often a flawed way of thinking. Regardless of %, big threads may be still as much important as medium and smaller threads, and sometimes even more important, because they cover big and popular websites which have many users (as many as hundreds or thousands smaller and medium websites combined which are reported in separate threads):

- https://github.com/uBlockOrigin/uAssets/issues?q=is%3Aissue%20sort%3Acomments-desc
- https://github.com/uBlockOrigin/uAssets/discussions?discussions_q=%22issues+for+shorteners+%22+
- https://github.com/AdguardTeam/AdguardFilters/issues?q=is%3Aissue%20sort%3Acomments-desc

**Comment by glaebhoerl** [comment]
It would be nice if the Load More button were split in two, the way it already works for diffs, so that loading more of the older comments resp. more of the newer comments would be separate actions.

This is *complementary* to improving the workflow for invoking Load More *multiple* times (whether by adding a Load All option, or at least avoiding have to chase the button across the page after each click).

**Comment by jab** [comment]
Even if you link directly to a comment, GitHub still hides it behind a "Load More". Example: https://github.com/jab/bidict/pull/242#discussion_r824305794

This link is obtained by choosing "Copy link" from the ••• menu for the associated comment. But if you open this link, you're taken to the top of the page, without the corresponding comment getting automatically expanded and scrolled to. Instead, it's as if you've been linked to a comment that does not exist. You need to somehow know to click the correct "Load more" link somewhere on the page and then somehow know which of the many newly-shown comments was being linked to. (In this case, the sole comment with the word "prefixed".)

This seems like too terrible a UX for this issue to have gone unfixed for so long, so I'm wondering if the GitHub team isn't actually tracking this issue internally. Should this be reported any other way?

**Comment by lostest4352** [comment]
Yeah its a terrible UX and very poorly thought out. There isn't even pages to jump to. You can't even select if "load more" loads the newer comments. Its always older comments.

**Comment by thw0rted** [comment]
I thought we already solved the UX problem of displaying a large number of items on a web page.  What's wrong with pagination?  User preference to save comments-per-page, maybe a separate one for whether to start with oldest vs newest; boom, done.  The current "Load More" behavior is almost never what the user actually wants to do.

**Comment by akpmohan07** [comment]
I am using https://marketplace.visualstudio.com/items?itemName=GitHub.vscode-pull-request-github

**Comment by garry-ut99** [comment]
Info: since the current thread, there have been two duplicates created:
- [Proposal: Allow loading more comments in either direction #10386](https://github.com/orgs/community/discussions/10386)
- [Improve "Load more" functionality for long PRs](https://github.com/orgs/community/discussions/17446)

And now some funny example as a proof of "github collapsed comments issue" being pain in ass in big threads:
- https://github.com/uBlockOrigin/uAssets/issues/20586#issuecomment-2157315282 :

> AutistOctavius : I was asked by u/paintboth1234 on Reddit to discuss my YouTube issues here as well as on [Reddit](https://www.reddit.com/r/uBlockOrigin/comments/1atwzem/comment/l7vdsg6/). I notice there are around 1200 posts in this thread, is there a way for me to read just ones from the last couple days without opening all 1200 posts?

- https://github.com/uBlockOrigin/uAssets/issues/20586#issuecomment-2161214094 :
> ComfortableMilk4454 : does anyone know how to expand all of the 1,317 hidden items? if i left click it it only expands a few more and id have to click like 100 times to see the whole conversation. trying to read through @ stephenhawk8054's recent message that contains the fix for the new in video yt ads. thanks so much!

Due to it, in big threads users are limited to read comments only from last 12 hours, it must be a joke from GitHub, the solution is simple: "Load more" button should be able also to load the last comments from the bottom, and not only from the top.

Despite the issue exists since 3 years already, still nothing has been done with it, it seems ignoring is how GitHub solves the problem.

@ AutistOctavius  @ComfortableMilk4454

**Reply by AutistOctavius** [reply]
> Info: since the current thread, there have been two duplicates created:
> 
> * [Proposal: Allow loading more comments in either direction #10386](https://github.com/orgs/community/discussions/10386)
> * [Improve "Load more" functionality for long PRs](https://github.com/orgs/community/discussions/17446)
> 
> And now some funny example as a proof of "github collapsed comments issue" being pain in ass in big threads:
> 
> * [🟥 ALL YouTube issues: D-O-N-'T COMMENT without reading the 1st post🟥 uBlockOrigin/uAssets#20586 (comment)](https://github.com/uBlockOrigin/uAssets/issues/20586#issuecomment-2157315282)
> 
> > I was asked by u/paintboth1234 on Reddit to discuss my YouTube issues here as well as on [Reddit](https://www.reddit.com/r/uBlockOrigin/comments/1atwzem/comment/l7vdsg6/). I notice there are around 1200 posts in this thread, is there a way for me to read just ones from the last couple days without opening all 1200 posts?
> 
> Due to it, in big threads users are limited to read comments only from last 12 hours, it must be a joke from GitHub, the solution is simple: "Load more" button should be able also to load the last comments from the bottom, and not only from the top.
> 
> @AutistOctavius

Yeah, it's a pain. Almost as much of a pain as YouTube and their insistence that I watch their SSAP ads.

**Comment by dm17** [comment]
Is there a way to "load all" yet? It takes me significantly less time to find this discussion and post this comment than it does to click "load earlier comments" on some long threads!

**Comment by tats-u** [comment]
This annoying "hidden items" prevents us from searching a word in all comments.

**Comment by RmStorm** [comment]
This discussion is now more than three years old and nothing at all has been done to improve it. I understand not many issues reach the size where this is an annoyance but for the issues that do it is a real hassle. Can someone in Github please prioritize this? It seems like a straightforward thing to at least improve like @glaebhoerl 's suggestion [here](https://github.com/orgs/community/discussions/5119#discussioncomment-1825803): `Load above / Load below`.

Introducing pagination or something might be more proper and take a bunch more work but that's not required to drastically improve the situation!

**Comment by HaloFour** [comment]
I think this is particularly bad in discussions.  I don't know what the solution could be, but better ways to sort/filter/search comments on a discussion would be helpful.  Not to mention, a link to a comment should Just Work ™️ .   Right now someone could response to a thread and you get that notification email, click on the link, and it only takes you to the top of the discussion because that one comment happened to be in response to a thread that was outside of the range loaded by default.  Sorting by newest doesn't help because that doesn't include replies.  At best you can search for the "Load more" link over and over again to load in all of the comments, and then try to search for some text in the reply.

**Comment by ppershing** [comment]
Another super-annoying part is that clicking on the Load More jumps the content to the beginning of the issue so now I have to re-scan all previous comments just to figure out where I should continue reading. This makes me very unproductive because it breaks my mental context before I find the place to continue.

**Reply by dm17** [reply]
100% I'll never understand how "UX experts" thought this was a good idea

**Reply by mootari** [reply]
It's a regression/bug. I'm guessing that they've changed their rendering to replace all existing comments and apparently didn't consider that they need to maintain or restore the scroll offset.

Edit: Here's a bug report someone else already filed: https://github.com/orgs/community/discussions/172523

**Comment by garry-ut99** [comment]
They've finally implemented it, pressing a 3-dot button on the right side gives two choices: _"Load older activity"_ and _"Load newer activity"_, it uncollapses 100-150 comments per a click:

<img width="1461" height="335" alt="111" src="https://github.com/user-attachments/assets/c9af7b45-9f55-44c0-8a48-9920203fd167" />

**Reply by Skgland** [reply]
Looks like it has only been implemented for Issues:

[<img width="1101" height="172" alt="image" src="https://github.com/user-attachments/assets/44d19288-499a-453a-b611-93e6deef1c1b" />](https://www.github.com/rust-lang/rust/issues/34511)

as for PRs I still get:

[<img width="894" height="96" alt="image" src="https://github.com/user-attachments/assets/428d0cc0-81ab-4b1b-9b30-165784f4f81c" />](https://www.github.com/rust-lang/rust/pull/112049)

without a menu to choose a direction.

(Though I was under the impression I had also seen it for PRs in the past but I am unable to reproduce that now)

**Reply by iandunn** [reply]
That helps, but it's not that much of an improvement. It's 2 clicks to get there, and and I still have to do it multiple times to load more comments.

There's already the "load more" button for loading a small amount at a time. The new button should just load them all. 

Maybe they're worried that folks won't realize that loading 2,500 comments might take a while, and could slow down or crash browsers on low-powered devices? I think that potential problem will be pretty obvious to most GitHub users. Those of us who are looking to show all comments are more likely to be maintainers, and are even more likely than the average user to be aware of potential issues.

**Reply by tats-u** [reply]
Insufficient. It works only once. The second banner does not have the ⋯ icon.

**Reply by garry-ut99** [reply]
Interesting indeed and weird, however it doesn't disappear always, sometimes the `⋯` button disappears, but sometimes not, looks like a random bug, or maybe there is some pattern...I've noticed that as long as _"Load older activity"_ or _"Load newer activity"_ buttons are used without using the main _"Load more"_ button , the `⋯` button is always present, but once you begin mixing them with the main _"Load more"_ button, and clicking all of them alternately, the `⋯` button sometimes disappear. The issue occurs on both Chromium & Firefox.

**Comment by StudioNirin** [comment]
The issue is that you still have to scroll to the bottom of the page to get to the load more... which then scrolls you all the way back to the top of the page again, so you have to scroll back down to the bottom to get to the items that are actually important. 

Who designed this UI? Was it vibe coded?

**Reply by garry-ut99** [reply]
Yes, the issue already mentioned in the first comment still does exist, it's awful and burdensome, they should implement a nice solution like on the flarum forum: https://discuss.flarum.org/d/32901-staff-diary-v20-cycle -   a nice slidebar at the right side which dynamically loads a given bunch of comments easily and fast. GitHub is 10 years behind...

**Reply by dm17** [reply]
10 years behind what? What is this future people are living in where they don't want information loaded and searchable? Why force the user to search via server side functions when their computers are powerful enough to search for a word on the page that is beyond the immediate scroll position? I don't understand how this is not "bad on purpose". Is this some kind of climate activism type thing - but they won't say it? Where loading the rest of the page when you didn't look at it is costing energy? Then just have a "load all" button - unless even the user intention to want to get all of the information should be opposed, limited or handicapped?

**Reply by garry-ut99** [reply]
See: https://github.com/orgs/community/discussions/5119#discussioncomment-15434232

**Reply by HaloFour** [reply]
A separate server-side search that is scoped to within the current discussion/issue/PR would be nice.

Also, if this discussion had enough comments, that link to a specific comment would simply not work, which is a terrible experience.

**Comment by aryanthopate** [comment]
GitHub should provide a “Load all comments” option to fully expand an issue or PR thread in one action. At minimum, when “Load more” is clicked, the page should automatically scroll to the end of the newly loaded comments so users can continue expanding without manually searching. This would significantly improve usability, especially when trying to view the most recent hidden comments.

**Reply by Skgland** [reply]
I don't think scrolling is the right solution. When I am searching for a comment I want to either search from the beginning forward or from the end backwards starting at the point the load more button was at.

So if it moved me to the new load more button I would have to scroll back to search the comments and then scroll forward again if it wasn't in the expanded section. If instead it stayed at the position of the expanded load more I would only have to scroll forward searching the comments while I move to the new load more button.

**Reply by tats-u** [reply]
You may want to check or search only recent comments.

**Reply by Skgland** [reply]
> You may want to check or search only recent comments.

Yes, I would scroll down to the end then search backwards. When reaching load more, I would click `load newer activity` and continue scrolling up.

[<img width="1101" height="172" alt="image" src="https://github.com/user-attachments/assets/44d19288-499a-453a-b611-93e6deef1c1b" />](https://www.github.com/rust-lang/rust/issues/34511)


```
Already Expanded history prefix

Load More newer/older activity buttons

Already Expanded history suffix
```

Expectation after clicking load more older activity

```
Already Expanded history prefix

Scroll position (Load more button used to be here)

Newly Expanded history

Load More newer/older activity buttons

Already Expanded history suffix
```

Expectation after clicking load more newer activity

```
Already Expanded history prefix

Load More newer/older activity buttons

Newly Expanded history

Scroll position (Load more button used to be here)

Already Expanded history suffix
```

Basically the expanded history should always be between my scroll position and the new position of the load more newer/older activity buttons.

Currently it appears to always mess up my scroll position both when loading newer activity and when loading older activity.

**Reply by garry-ut99** [reply]
> aryanthopate: GitHub should provide a “Load all comments” option to fully expand an issue or PR thread in one action

Bad / risky idea, a duplicate of: https://github.com/orgs/community/discussions/5119#discussioncomment-1584228, already discussed and not considered a good solution for big threads.

**Reply by tats-u** [reply]
> Bad / risky idea

If you think so how can we search for a specific keyword in all comments?

**Reply by garry-ut99** [reply]
Well, it's an another / different problem, perhaps you should ask github, not me, as it seems to be github's design flow / side effect of their design choice. I don't say you can't uncollapse all comments, but only that doing it in a single run is risky, as unrolling several hundreds or thousands comments at once is a recipe for a lag / server overload or unresponsivness, connection drop, browser crash or a frozen tab, etc. Each comment isn't just text; it's a collection of profile pictures, Markdown-rendered formatting, reaction emojis, and event listeners. If one wants to avoid the performance "death spiral" of a massive DOM tree, it's not advised to uncollapse massive threads in a single run. Even on most of typical forums (phpBB etc) there are pages and each page contains only 20-50 comments. Perhaps Github should implement "Search within this thread" feature in the web UI, so that only comments which contain the desired keyword are uncollapsed, perhaps altogether with a bunch of surrounding comments (context).

**Reply by andry81** [reply]
May be this can help: https://github.com/ludios/expand-everything

**Reply by garry-ut99** [reply]
We're rather discussing about a native implementation, and we are aware of userscripts like mentioned already in: https://github.com/orgs/community/discussions/5119#discussioncomment-1641624, anyway, userscript you've provided is outdated like most of them and doesn't seem to work anymore (at least for me in Firefox), did you ever test it before recommending?

**Reply by thw0rted** [reply]
[As I said](https://github.com/orgs/community/discussions/5119#discussioncomment-8055979) (2 years ago tomorrow!)  there are multiple use cases and they're all solved UX problems -- at least for issue / PR comments, which are flat.

If the user wants to just read through comments in chronological order, pagination is easy.  If they want to search for keywords, just add a "search this issue/PR" to the existing top level search bar.

Why are we still debating this?

**Reply by dm17** [reply]
Ya it is bad on purpose. I think I'll just use a script like suggested above when I run into this instead of expecting these companies to change their ethos:

> May be this can help: https://github.com/ludios/expand-everything

**Reply by andry81** [reply]
@garry-ut99 
> anyway, userscript you've provided is outdated like most of them and doesn't seem to work anymore (at least for me in Firefox), did you ever test it before recommending?

It works at least for discussions. You have to enable the script and click the reload button.

**Reply by andry81** [reply]
@thw0rted 
> pagination is easy

Pagination suck at expand-all-and-save-in-a-single-file.

**Reply by garry-ut99** [reply]
> andry81: It works at least for discussions. You have to enable the script and click the reload button.

But still doesn't work for _Issues_, so it works for _Discussions_, but not for _Issues_. 

I didn't test it on _Discussions_ before, because after it didn't work for _Issues_ I assumed it doesn't work anywhere, as I expected it to work everywhere or nowhere. Not working on _Issues_, while still working on _Discussions_, is a bit unexpected and inconsistent, and it seems that you were misled by the same logic as well: you tried it on _Discussions_, and wrongly assumed it works on _Issues_. The userscript fooled us both.

<br>

> andry81: You have to enable the script and click the reload button.

I know, you have to do the same to realise it doesn't work on _Issues._

**Reply by andry81** [reply]
@garry-ut99 
> you tried it on _Discussions_, and wrongly assumed it works on _Issues_. The userscript fooled us both.

That's why I said "may be this can help".

**Reply by andry81** [reply]
@garry-ut99 
> I know, you have to do the same to realise it doesn't work on _Issues._

You can try a fix: https://github.com/ludios/expand-everything/pull/12

**Reply by garry-ut99** [reply]
Yes, it works now.

**Reply by thw0rted** [reply]
> expand-all-and-save-in-a-single-file

This could be addressed with a "print" button (to make an all-comments PDF) or using the REST API.

**Comment by MrSuddenJoy** [comment]
If I were GitHub, I would implement `lazy-loading`. Thats it.

**Comment by jcsahnwaldt** [comment]
Terrible UX. As a workaround, there's a JS script that loads all comments: https://gist.github.com/jamesob/63b1b92ad3d1c6bbddf2677073a087fb
